core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalReadConverterContext.scala

```diff
@@ -17,59 +17,127 @@
+
+  private def plainString(chunk: Chunk) = chunk.atStr(new BufferedString(), rowIdx).toString
+
+  val StringProviders = Map[Byte, (Chunk => String)](
+    Vec.T_CAT -> categoricalString,
+    Vec.T_UUID -> uuidString,
+    Vec.T_STR -> plainString
+  ) withDefault((t:Byte) => {
+    assert(assertion = false, s"Should never be here, type is $t")
+    (_:Chunk) => null
+  }
+    )
+
+  def string(chunk: Chunk): String = StringProviders(chunk.vec().get_type())(chunk)
+
+  // TODO(vlad): check if instead of stringification, we could use bytes
+  private def utfString(chunk: Chunk) = UTF8String.fromString(string(chunk))
 
   private def underlyingFrame = DKV.get(Key.make(keyName)).get.asInstanceOf[Frame]
 
-  override def numRows: Int = nrows
+  private val DefaultReader: Reader = (_:Int) => None
+
+  import SupportedTypes._
+  /**
+    * Given a a column number, returns an Option[T]
+    * with the value parsed according to the type.
+    * You can override it.
+    *
+    * A map from type name to option reader
+    */
+  private lazy val ExtractorsTable: Map[SimpleType[_], Chunk => _] = Map(
+    Boolean    -> booleanAt _,
+    Byte       -> byteAt _,
+    Double     -> doubleAt _,
+    Float      -> floatAt _,
+    Integer    -> intAt _,
+    Long       -> longAt _,
+    Short      -> shortAt _,
+    String     -> plainString _,
+    Category   -> categoricalString _,
+    UUID       -> uuidString _,
+    UTF8       -> utfString _,
+    Timestamp  -> timestamp _
+  )
+
+  private lazy val OptionReadersMap: Map[OptionalType[_], OptionReader] =
+    ExtractorsTable map {
+      case (t, getter) => SupportedTypes.byBaseType(t) -> returnOption(getter) _
+    } toMap
+
+  private lazy val OptionReaders: Map[OptionalType[_], OptionReader] = OptionReadersMap withDefault
+    (t => throw new scala.IllegalArgumentException(s"Type $t conversion is not supported in Sparkling Water"))
+
+  private lazy val SimpleReadersMap: Map[SimpleType[_], Reader] =
+    ExtractorsTable map {
+      case (t, getter) => t -> returnSimple(t.defaultValue, getter) _
+    } toMap
+
+  private lazy val SimpleReaders: Map[SimpleType[_], Reader] = SimpleReadersMap withDefault
+    (t => throw new scala.IllegalArgumentException(s"Type $t conversion is not supported in Sparkling Water"))
+
+  private def plainReaderFor(supportedType: SimpleType[_], optReader: OptionReader): Reader =
+    (col: Int) => optReader(col).getOrElse(supportedType.defaultValue)
+
+  lazy val readerMapByName: Map[NameOfType, Reader] = (OptionReaders ++ SimpleReaders) map {
+    case (supportedType, reader) => supportedType.name -> reader
+  } toMap
+
+  def columnValueProviders(columnIndexesWithTypes: Array[(Int, SimpleType[_])]): Array[() => Option[Any]] = {
+    for {
+      (columIndex, supportedType) <- columnIndexesWithTypes
+      reader = OptionReaders(byBaseType(supportedType))
+      provider = () => reader.apply(columIndex)
+    } yield provider
+  }
 }
```

Comment:
Well, now (at last) I see what "Internal" means here. :) Yes, we can move it over there, or, I would suggest, to a separate trait that both Internal and External extend. I could do it right now, but it's not exactly clear to me which parts are worth moving. All except ExtractorMap? If so, that would be totally gorgeous. OTOH, I would not rush with this; as soon as External starts forming, it will be much easier to figure out which parts are actually worth sharing.